#19 Add quota support

raymondberg opened the pull request:
__Problem__

In mid-2020, Redshift added [quota
support](https://aws.amazon.com/about-aws/whats-new/2020/06/announcing-storage-controls-for-schemas-amazon-redshift/)
to limit the size of schemas. This feature is not yet supported in the
provider. Full details in #19

__Solution__

@coopergillan and I spiked a basic capability. After a few hours of
debugging I verified the code can perform both creates and updates using
quotas. Docs included in solution

The key design idea was that unlimited quotas are specified as strings,
whereas all other quotas are specified as numbers. To get around this,
we can use a fairly common standard of saying 0 = unlimited. There
appears to be no conflicts with other behaviors since 0 is not supported
on Redshift today due to a (at time of authorship) 2048MB lower-limit.
That limit seems to imply that it would be unlikely AWS would ever add a
"0 MB" limit that behaved differently than unlimited.

While GB is the default unit of measure in the documentation, getting
granularity of MB seems to cover all use-cases without having to do
string validation on a string-type parameter: e.g. '25 GB'. That said,
I'm open to the explicit string being a better use-case.

raymondberg commented:
Note that, as a result of using quotas, I added the new requirements to the README. This includes a new base-supported branch for Redshift (since the reads now require quota support). This would also make a breaking change for folks who wanted to get on 1.x without upgrading redshift clusters. While rare to run older versions, I think we should stick to [Semantic Versioning](https://semver.org/) and bump any new releases to 2.x.

raymondberg commented:
Lastly, I looked into adding [LocalStack](https://github.com/localstack/localstack) to support automated tests of the API, but it appears [table-level Redshift is not yet supported](https://github.com/localstack/localstack/issues/833) on the community edition. I don't know they would supply us a license for this open-source project.

coopergillan commented on redshift/resource_redshift_schema.go:
```diff
@@ -42,6 +43,12 @@ func redshiftSchema() *schema.Resource {
+				Type:        schema.TypeInt,
+				Optional:    true,
+				Description: "In megabytes, the maximum amount of disk space that the specified schema can use",
+				Default:     0,
```
I am open to leaving this as-is, but I wonder if it would be better to provide a non-unlimited value by default? I suppose this is the pre-quota definitions case though.

raymondberg replied:
Given that this was the behavior before the addition, it seemed reasonable.

coopergillan replied:
Ah, good point.

coopergillan commented on redshift/resource_redshift_schema.go:
```diff
@@ -120,9 +134,13 @@ func readRedshiftSchema(d *schema.ResourceData, db *sql.DB) error {
 
-	err := db.QueryRow("select nspname, nspowner from pg_namespace where oid = $1", d.Id()).Scan(&schemaName, &owner)
+	err := db.QueryRow(`
+			select trim(nspname) as nspname, nspowner, coalesce(quota, 0) as quota from pg_namespace
```
Super nitpick: could we move `from pg_namespace` to the next line so it's just a little more clear how this query is structured?

raymondberg replied:
c50e1eb04

coopergillan reviewed (approved):
A couple non-breaking questions. This is awesome!

👟

coopergillan commented:
@raymondberg - one other doc-related suggestion: let's add a quota in [the example for creating a schema](https://github.com/coopergillan/terraform-provider-redshift/blob/38d67e316a5aaf3a46afedab115c832e4d595f48/example/root.tf#L40-L43).

coopergillan commented:
@raymondberg - just merged https://github.com/coopergillan/terraform-provider-redshift/pull/25 so we could add the quota example in the newly-formatted root.tf.

raymondberg committed:
query restyle

raymondberg committed:
Add support for schema quotas  (addresses #19)

__Problem__

In mid-2020, Redshift added [quota
support](https://aws.amazon.com/about-aws/whats-new/2020/06/announcing-storage-controls-for-schemas-amazon-redshift/)
to limit the size of schemas. This feature is not yet supported in the
provider. Full details in #19

__Solution__

@coopergillan and I spiked a basic capability. After a few hours of
debugging I verified the code can perform both creates and updates using
quotas. Docs included in solution

The key design idea was that unlimited quotas are specified as strings,
whereas all other quotas are specified as numbers. To get around this,
we can use a fairly common standard of saying 0 = unlimited. There
appears to be no conflicts with other behaviors since 0 is not supported
on Redshift today due to a (at time of authorship) 2048MB lower-limit.
That limit seems to imply that it would be unlikely AWS would ever add a
"0 MB" limit that behaved differently than unlimited.

While GB is the default unit of measure in the documentation, getting
granularity of MB seems to cover all use-cases without having to do
string validation on a string-type parameter: e.g. '25 GB'. That said,
I'm open to the explicit string being a better use-case.

raymondberg committed:
Add quota limit example and cleanup root.tf

raymondberg commented:
Had to do some rebasing, but it should be good now

coopergillan reviewed (approved):
👟
